### ISSUE TITLE
old build files not cleaned up automatically in dev mode

### ISSUE THREAD

**Issue by akiran** [remix-run/remix#2432]
### What version of Remix are you using?

1.1.3

### Steps to Reproduce

Create a project with remix and continue to make changes in the projects. New build files with new hashes are generated on every code change. Build folder size continues to increase drastically over time as the old build files are not cleaned up.

### Expected Behavior

Old and unused build files should be automatically removed to avoid a drastic increase in the build folder size

### Actual Behavior

In #2425 discussion, @sergiodxa mentioned that this issue was fixed in #252
I tried stopping the remix dev server and re-started it. But the build folder is not cleaned up automatically.

Currently, my build folder size is 24 GB.

**Comment by jelofsson** [comment]
I have the same behavior, build folder is not cleaned up automatically.

**Comment by axel-xitaso** [comment]
I have encountered the same issue. My workaround is to remove the two Remix output directories before starting the dev server.

**Comment by kelleyperry** [comment]
<img width="404" alt="Screen Shot 2022-04-02 at 11 25 03 AM" src="https://user-images.githubusercontent.com/4185768/161396378-8470ab76-ae1b-4249-a1c0-83c1f4c182dc.png">

**Comment by haach** [comment]
I am running into the same issue. Just cleaned out almost 10GB of files. 😱 Was there any progress on this?

**Comment by thomasverleye** [comment]
For the contributors who would start on this, I did some testing and came up with following observation I posted in my duplicate ticket:

> Been testing the watch function from the CLI Commands.
> It seems like the `exitHook` can't be triggered properly.
> See: https://github.com/remix-run/remix/blob/main/packages/remix-dev/cli/commands.ts#L222
> 
> When testing the `exitHook` inside the dev function, it does get triggered, so I'm afraid the watch function can't be run async and maybe should return a callback function instead closes everything correctly inside the dev function.
> 
> Not sure if that's a solid solution though.

**Comment by dmarkow** [comment]
The issue is that `exit-hook` package doesn't support async code ([yet](https://github.com/sindresorhus/exit-hook/pull/20)), so as soon as `resolve` is called the process exits without waiting for the `then` block to execute.  https://github.com/remix-run/remix/blob/15213866f8dca94e6164c46028f410625226599f/packages/remix-dev/cli/commands.ts#L226-L237

**Comment by eladchen** [comment]
@dmarkow async is now supported by the exit-hook package.

**Comment by qaynam** [comment]
any solution?

<img width="392" alt="image" src="https://user-images.githubusercontent.com/33139713/197205450-9e68a41b-cf10-430b-941c-3e3777dc58d5.png">

**Comment by kiliman** [comment]
I just add a "clean" script that runs before "dev" that deletes the build folders.
```json
"clean": "rimraf build public/build",
"dev": "npm run clean && remix dev",
```
You'll need to add the `rimraf` package

**Comment by qaynam** [comment]
> I just add a "clean" script that runs before "dev" that deletes the build folders.
> 
> ```json
> "clean": "rimraf build public/build",
> "dev": "npm run clean && remix dev",
> ```
> 
> You'll need to add the `rimraf` package

but what about hot reload ?
problem is every time source file got changed , remix bundler regenerate new version of JS(or css) file , but not clear build dir, it build different file to build dir over and over again.

**Comment by kiliman** [comment]
True. This is primarily so my build folders are not unbounded.

**Comment by eladchen** [comment]
This issue can be solved by using esbuild in-memory ("write: false" option).

https://github.com/evanw/esbuild/issues/139
https://github.com/evanw/esbuild/commit/268f8331f5b3840e659c600b9ddf797b3a147018

When content is served by the remix request handlers, remix can lookup the resource within the in memory compilation
and in case one is not found look for the resource in the file system...

Using an in-memory would most likely result in faster compilations.

WDYT?

**Comment by universse** [comment]
not sure what the downsides are but what if the hashing of file names is disabled in dev mode?

**Comment by pcattori** [comment]
Fixed by #5223
